### PR TITLE
HPCC-28118 Add ws_access.AccountPermissionsV2

### PR DESCRIPTION
**Pull request by wangkx**
The ws_access.AccountPermissionsV2 may be used to replace the existing ws_access.AccountPermissions. The existing ws_access.AccountPermissions has several duplicated calls to LDAP server. In this version, they are removed.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by wangkx** [comment]
@RussWhitehead please also verify if there is any missing function from ws_access.AccountPermissions.

**Review by RussWhitehead — APPROVED** [review]
@wangkx Looks good. How can we get this integrated into ECLWatch ?

**Comment by wangkx** [comment]
@ghalliday please review and merge.

**Comment by wangkx** [comment]
> How can we get this integrated into ECLWatch ?

https://track.hpccsystems.com/browse/HPCC-27685

**Comment by ghalliday** [comment]
@wangkx should this be targeting 8.12.x instead?  Any significant reason why it should target 8.10.x?

File: esp/services/ws_access/ws_accessService.cpp [hunk 1]
```diff
@@ -4009,6 +4025,467 @@ bool Cws_accessEx::onAccountPermissions(IEspContext &context, IEspAccountPermiss
...
+        StringBuffer userID;
+        context.getUserID(userID);
+        if (userID.isEmpty())
+            throw makeStringException(ECLWATCH_INVALID_INPUT, "Could not get user ID.");
+
+        const char *accountName = req.getAccountName();
+        if (!isEmptyString(accountName) && !streq(accountName, userID.str()))
+            checkUser(context, secMGR);
+
+        //Make sure BaseDN settings loadded
```
Comment (ghalliday) [inline esp/services/ws_access/ws_accessService.cpp:4475]: trivial: loaded

File: esp/services/ws_access/ws_accessService.cpp [hunk 2]
```diff
@@ -4009,6 +4025,467 @@ bool Cws_accessEx::onAccountPermissions(IEspContext &context, IEspAccountPermiss
...
+                        //Not find the account in this resource. Add the BaseDN name to the basednNames.
+                        basednNames.append(accountsInBaseDN.getBaseDNName());
+                        break;
+                    }
+                }
+            }
+        }
+
+    public:
+        CAccountPermissionsHelper(IEspContext *ctx, CLdapSecManager *secmgr) : context(ctx), secMGR(secmgr) { };
```
Comment (ghalliday) [inline esp/services/ws_access/ws_accessService.cpp:4394]: trivial: trailling ; is unnecessary

File: esp/services/ws_access/ws_accessService.cpp [hunk 3]
```diff
@@ -4009,6 +4025,467 @@ bool Cws_accessEx::onAccountPermissions(IEspContext &context, IEspAccountPermiss
...
+//Revised based on onAccountPermissions() which lists permissions for a given account in all BaseDN resources.
+bool Cws_accessEx::onAccountPermissionsV2(IEspContext &context, IEspAccountPermissionsV2Request &req,
+    IEspAccountPermissionsV2Response &resp)
+{
+    class CAccountsInResource : public CInterface
+    {
+        StringAttr resourceName;
+        StringArray accountNames;
+    public:
+        CAccountsInResource(const char *_resourceName) : resourceName(_resourceName) {};
```
Comment (ghalliday) [inline esp/services/ws_access/ws_accessService.cpp:4038]: trivial: a few spurious ; after the inline function definitions.

File: esp/services/ws_access/ws_accessService.cpp [hunk 4]
```diff
@@ -4009,6 +4025,467 @@ bool Cws_accessEx::onAccountPermissions(IEspContext &context, IEspAccountPermiss
...
+                permission->setDeny_access(true);
+            if((denies & NewSecAccess_Read) == NewSecAccess_Read)
+                permission->setDeny_read(true);
+            if((denies & NewSecAccess_Write) == NewSecAccess_Write)
+                permission->setDeny_write(true);
+            if((denies & NewSecAccess_Full) == NewSecAccess_Full)
+                permission->setDeny_full(true);
+            return permission.getClear();
+        }
+        void appendAccountPermission(Owned<IEspAccountPermission> &permissionToBeAppended,
```
Comment (ghalliday) [inline esp/services/ws_access/ws_accessService.cpp:4198]: Owned<IEspAccountPermission> &permissionToBeAppended,

This is an unusual type for a parameter.  It is generally only used if the parameter is going to be modified.  If it isn't then either

const Owned<IEspAccountPermission> &permissionToBeAppended,

or more commonly

IEspAccountPermission * permissionToBeAppended,

and LINK(permissionToBeAppended)

File: esp/services/ws_access/ws_accessService.cpp [hunk 5]
```diff
@@ -4009,6 +4025,467 @@ bool Cws_accessEx::onAccountPermissions(IEspContext &context, IEspAccountPermiss
...
+            {
+                IEspGroupAccountPermission &groupPermission = groupAccountPermissions.item(i);
+                if (!streq(actName, groupPermission.getGroupName()))
+                    continue;
+
+                //This perm is for a group account which is already in the groupPermission.
+                //Append the Account Permission into the groupPermission.
+                IArrayOf<IConstAccountPermission> &permissions = groupPermission.getPermissions();
+                permissions.append(*permissionToBeAppended.getLink());
+                appended = true;
```
Comment (ghalliday) [inline esp/services/ws_access/ws_accessService.cpp:4245]: trivial: more consistent with the other branches above would be "return;"

**Review by ghalliday — COMMENTED** [review]
@wangkx a few minor style comments - none would prevent it being merged.  I defer to Russ's knowledge whether the code is correct.  Main question is which build should this target.

**Comment by wangkx** [comment]
@RussWhitehead as I know, OPS may need this sooner. But, this is a new. Should it be targeting to 8.12.x?
@ghalliday I made the changes based on your review.

**Comment by RussWhitehead** [comment]
@wangkx Ops will not benefit from this feature until ECLWatch is modified to call it. I recommend opening a subtask Jira for ECLWatch, and retargeting to 8.12.x

**Comment by wangkx** [comment]
@ghalliday this PR is re-targeted to 8.12.x.

**Review by ghalliday — APPROVED** [review]
@wangkx looks good, please squash

**Comment by wangkx** [comment]
Squashed. @ghalliday please merge.

**Pull request merged**